java/driver/jni/src/main/java/org/apache/arrow/adbc/driver/jni/impl/NativeAdbcException.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.adbc.driver.jni.impl;
+
+/**
+ * Wrapper for a {@code struct AdbcException} that doesn't depend on other Java code.
+ *
+ * <p>This breaks a dependency chain in compilation. If we directly used AdbcException, then we
+ * would need to compile the Java code before compiling JNI code, but we need the compiled JNI code
```

Comment:
this is surprising? I thought `javah` would use the source code, and that `javac` has no way to check native code? Is the circular dependency actually because of tests?

Reply:
The CMake code has to run javac to generate the JNI headers. And CMake doesn't resolve Maven dependencies. So it's easiest if these files can be compiled standalone.

Reply:
I can try to configure Maven to do this instead? It'd make the build a bit more complicated (run maven, then cmake, then maven) but maybe that's OK.

Reply:
Would invoking Maven from CMake to generate JNI headers address this?

Maybe using `native-maven-plugin:javah` goal?

Reply:
I could also do that. I sorta dislike invoking one build system from another as it makes it a bit more annoying to get in between them to debug or customize things but just generating headers should be OK.

Reply:
Updated to invoke Maven from CMake to generate headers, though I still need to rework the rest of the code now that we don't have the circular dependency

Reply:
We also use it to find the JNI headers, discover and invoke the C++ compiler, etc.